// vdl-angular/data-table/data-table.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component, Input, Output, EventEmitter, ContentChild, ContentChildren, QueryList, NgModule, ViewEncapsulation, ChangeDetectionStrategy } from '@angular/core';
import { VdlCheckboxModule } from '../checkbox/checkbox';
import 'rxjs/add/operator/share';
import { VdlDataTableHeaderSelectableRow, VdlDataTableSelectableRow } from './data-table-selectable-tr';
export var VdlDataTable = (function () {
    function VdlDataTable() {
        this.onSelectableAll = new EventEmitter(false);
        this.onSelectableChange = new EventEmitter(false);
        this._listeners = [];
        this.selected = [];
        this.onSelectableChange.share();
    }
    VdlDataTable.prototype.change = function (event) {
        var outputEvent = {
            name: 'selectable_change',
            allSelected: false,
            values: []
        };
        if (event.target instanceof VdlDataTableHeaderSelectableRow === true) {
            if (event.isActive === true) {
                outputEvent.allSelected = true;
                outputEvent.values = this._getRowsValues();
            }
        }
        else {
            outputEvent.values = this.selected.slice(0);
            if (event.isActive === true) {
                outputEvent.values.push(event.selectableValue);
            }
            else {
                var index = outputEvent.values.indexOf(event.selectableValue);
                if (index !== -1) {
                    outputEvent.values.splice(index, 1);
                }
            }
        }
        // dispatch change
        this.selected = outputEvent.values;
        this.onSelectableChange.emit(outputEvent);
    };
    /**
     * @returns {Array<string>}
     */
    VdlDataTable.prototype._getRowsValues = function () {
        return this._rows.toArray()
            .map(function (tr) { return tr.selectableValue; });
    };
    VdlDataTable.prototype._unsubscribeChildren = function () {
        this.selected = [];
        if (this._listeners.length) {
            this._listeners.forEach(function (listener) {
                listener.unsubscribe();
            });
            this._listeners = [];
        }
    };
    VdlDataTable.prototype._updateChildrenListener = function (list) {
        var _this = this;
        this._unsubscribeChildren();
        list.toArray()
            .map(function (tr) {
            tr.onChange.subscribe(_this.change.bind(_this));
        });
    };
    VdlDataTable.prototype.ngAfterContentInit = function () {
        if (this.selectable === true) {
            if (!!this._masterRow) {
                this._masterRow.onChange.subscribe(this.change.bind(this));
            }
            this._rows.changes.subscribe(this._updateChildrenListener.bind(this));
            this._updateChildrenListener(this._rows);
        }
    };
    VdlDataTable.prototype.ngOnDestroy = function () {
        this._unsubscribeChildren();
    };
    __decorate([
        Input(), 
        __metadata('design:type', Boolean)
    ], VdlDataTable.prototype, "selectable", void 0);
    __decorate([
        Output(), 
        __metadata('design:type', EventEmitter)
    ], VdlDataTable.prototype, "onSelectableAll", void 0);
    __decorate([
        Output(), 
        __metadata('design:type', EventEmitter)
    ], VdlDataTable.prototype, "onSelectableChange", void 0);
    __decorate([
        ContentChild(VdlDataTableHeaderSelectableRow), 
        __metadata('design:type', VdlDataTableHeaderSelectableRow)
    ], VdlDataTable.prototype, "_masterRow", void 0);
    __decorate([
        ContentChildren(VdlDataTableSelectableRow, true), 
        __metadata('design:type', QueryList)
    ], VdlDataTable.prototype, "_rows", void 0);
    VdlDataTable = __decorate([
        Component({selector: 'vdl-data-table',
            template: "<ng-content></ng-content>",
            styles: [".vdl-data-table,vdl-data-table{display:table;border-spacing:0;border-collapse:collapse;font-size:14px;box-sizing:border-box;width:100%}.vdl-data-table vdl-checkbox,vdl-data-table vdl-checkbox{margin:0;width:18px;vertical-align:middle;padding-bottom:3px}.vdl-data-table tr,vdl-data-table tr{vertical-align:middle}.vdl-data-table tr:last-child,vdl-data-table tr:last-child{height:56px}.vdl-data-table td,.vdl-data-table th,vdl-data-table td,vdl-data-table th{padding:0 32px 0 24px;text-align:right}.vdl-data-table td:first-of-type,.vdl-data-table th:first-of-type,vdl-data-table td:first-of-type,vdl-data-table th:first-of-type{padding-left:24px}.vdl-data-table td:last-of-type,.vdl-data-table th:last-of-type,vdl-data-table td:last-of-type,vdl-data-table th:last-of-type{padding-right:24px}.vdl-data-table td.vdl-text-cell,.vdl-data-table th.vdl-text-cell,vdl-data-table td.vdl-text-cell,vdl-data-table th.vdl-text-cell{text-align:left}.vdl-data-table td.vdl-data-check-cell,.vdl-data-table th.vdl-data-check-cell,vdl-data-table td.vdl-data-check-cell,vdl-data-table th.vdl-data-check-cell{padding:0 0 0 24px;width:18px}.vdl-data-table th,vdl-data-table th{font-size:16px;font-weight:600;color:#414142;text-overflow:ellipsis;box-sizing:border-box}.vdl-data-table th.sortable,vdl-data-table th.sortable{cursor:pointer}.vdl-data-table th.sorted-ascending,.vdl-data-table th.sorted-descending,vdl-data-table th.sorted-ascending,vdl-data-table th.sorted-descending{color:#414142}.vdl-data-table th.sorted-ascending::before,.vdl-data-table th.sorted-descending::before,vdl-data-table th.sorted-ascending::before,vdl-data-table th.sorted-descending::before{font-family:'Material Icons';font-size:16px;content:'\e5d8';margin-right:5px;vertical-align:sub}.vdl-data-table th.sorted-ascending:hover,.vdl-data-table th.sorted-descending:hover,vdl-data-table th.sorted-ascending:hover,vdl-data-table th.sorted-descending:hover{cursor:pointer}.vdl-data-table th.sorted-ascending:hover::before,.vdl-data-table th.sorted-descending:hover::before,vdl-data-table th.sorted-ascending:hover::before,vdl-data-table th.sorted-descending:hover::before{color:rgba(0,0,0,.38)}.vdl-data-table th.sorted-descending::before,vdl-data-table th.sorted-descending::before{content:'\e5db'}.vdl-data-table td,vdl-data-table td{position:relative;vertical-align:middle;height:48px;border-top:1px solid rgba(0,0,0,.12);border-bottom:1px solid rgba(0,0,0,.12);box-sizing:border-box;color:#414142}.vdl-data-table td:first-child,.vdl-data-table th:first-child,vdl-data-table td:first-child,vdl-data-table th:first-child{padding-left:24px}.vdl-data-table td:last-child,.vdl-data-table th:last-child,vdl-data-table td:last-child,vdl-data-table th:last-child{padding-right:24px}.vdl-data-table tr,vdl-data-table tr{position:relative;height:48px}.vdl-data-table .vdl-data-tbody tr.active,.vdl-data-table .vdl-data-tbody tr.selected,.vdl-data-table tbody tr.active,.vdl-data-table tbody tr.selected,vdl-data-table .vdl-data-tbody tr.active,vdl-data-table .vdl-data-tbody tr.selected,vdl-data-table tbody tr.active,vdl-data-table tbody tr.selected{background-color:#e9f4fd}.vdl-data-table .vdl-data-tbody tr:hover,.vdl-data-table tbody tr:hover,vdl-data-table .vdl-data-tbody tr:hover,vdl-data-table tbody tr:hover{background-color:#e2ecf5}"],
            host: {
                '[class.vdl-data-table]': 'true',
                '[class.vdl-data-table-selectable]': 'selectable',
            },
            encapsulation: ViewEncapsulation.None,
            changeDetection: ChangeDetectionStrategy.OnPush,
        }), 
        __metadata('design:paramtypes', [])
    ], VdlDataTable);
    return VdlDataTable;
}());
export var VdlDataTableModule = (function () {
    function VdlDataTableModule() {
    }
    VdlDataTableModule.forRoot = function () {
        return {
            ngModule: VdlDataTableModule,
            providers: []
        };
    };
    VdlDataTableModule = __decorate([
        NgModule({
            declarations: [
                VdlDataTable,
                VdlDataTableHeaderSelectableRow,
                VdlDataTableSelectableRow
            ],
            exports: [
                VdlDataTable,
                VdlDataTableHeaderSelectableRow,
                VdlDataTableSelectableRow
            ],
            imports: [
                VdlCheckboxModule
            ]
        }), 
        __metadata('design:paramtypes', [])
    ], VdlDataTableModule);
    return VdlDataTableModule;
}());
