// vdl-angular/data-table/data-table-selectable-tr.js
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
import { Component, Output, Input, EventEmitter, Inject, Optional, forwardRef, ElementRef } from '@angular/core';
import 'rxjs/add/operator/share';
import 'rxjs/add/operator/map';
import { VdlDataTable } from './data-table';
export var AbstractVdlDataTableSelectableRow = (function () {
    function AbstractVdlDataTableSelectableRow(table, _element) {
        this.table = table;
        this._element = _element;
        this.onChange = new EventEmitter(false);
        this.isActive = false;
        this.onChange.share();
    }
    /**
     * Change active status
     */
    AbstractVdlDataTableSelectableRow.prototype.change = function (event) {
        if (event.target.classList.contains('vdl-checkbox-inner-container')) {
            event.preventDefault();
        }
        this.isActive = !this.isActive;
        var changeEvent = {
            name: 'selectable_row_change',
            target: this,
            isActive: this.isActive,
            selectableValue: this.selectableValue
        };
        this.onChange.emit(changeEvent);
    };
    AbstractVdlDataTableSelectableRow.prototype.ngAfterContentInit = function () { };
    __decorate([
        Input('selectable-value'), 
        __metadata('design:type', String)
    ], AbstractVdlDataTableSelectableRow.prototype, "selectableValue", void 0);
    __decorate([
        Output(), 
        __metadata('design:type', EventEmitter)
    ], AbstractVdlDataTableSelectableRow.prototype, "onChange", void 0);
    AbstractVdlDataTableSelectableRow = __decorate([
        __param(0, Optional()),
        __param(0, Inject(forwardRef(function () { return VdlDataTable; }))), 
        __metadata('design:paramtypes', [VdlDataTable, ElementRef])
    ], AbstractVdlDataTableSelectableRow);
    return AbstractVdlDataTableSelectableRow;
}());
export var VdlDataTableHeaderSelectableRow = (function (_super) {
    __extends(VdlDataTableHeaderSelectableRow, _super);
    function VdlDataTableHeaderSelectableRow(table, _element) {
        _super.call(this, table, _element);
        this.table = table;
        this._element = _element;
    }
    VdlDataTableHeaderSelectableRow.prototype._bindListener = function () {
        var _this = this;
        this.table.onSelectableChange
            .map(function (event) { return event.allSelected; })
            .subscribe(function (newActiveStatus) { return _this.isActive = newActiveStatus; });
    };
    VdlDataTableHeaderSelectableRow.prototype.ngAfterContentInit = function () {
        if (!!this.table) {
            this._bindListener();
        }
    };
    VdlDataTableHeaderSelectableRow = __decorate([
        Component({
            selector: 'tr[vdl-data-table-header-selectable-row]',
            template: "\n        <th class=\"vdl-data-check-cell\">\n            <vdl-checkbox [checked]=\"isActive\"></vdl-checkbox>\n        </th>\n        <ng-content></ng-content>\n    ",
            host: {
                '[class.active]': 'isActive',
                '(click)': 'change($event)'
            }
        }),
        __param(0, Optional()),
        __param(0, Inject(forwardRef(function () { return VdlDataTable; }))), 
        __metadata('design:paramtypes', [VdlDataTable, ElementRef])
    ], VdlDataTableHeaderSelectableRow);
    return VdlDataTableHeaderSelectableRow;
}(AbstractVdlDataTableSelectableRow));
export var VdlDataTableSelectableRow = (function (_super) {
    __extends(VdlDataTableSelectableRow, _super);
    function VdlDataTableSelectableRow(table, _element) {
        _super.call(this, table, _element);
        this.table = table;
        this._element = _element;
    }
    /**
     * @param {any} element
     *
     * @returns {string}
     */
    VdlDataTableSelectableRow.prototype._getIndex = function (element) {
        return Array.prototype.indexOf.call(element.parentNode.children, element).toString();
    };
    VdlDataTableSelectableRow.prototype._bindListener = function () {
        var _this = this;
        this.table.onSelectableChange
            .map(function (event) {
            return event.values !== undefined &&
                event.values.length &&
                (event.values.findIndex(function (value) { return value === _this.selectableValue; })) !== -1;
        })
            .subscribe(function (newActiveStatus) { return _this.isActive = newActiveStatus; });
    };
    VdlDataTableSelectableRow.prototype.ngAfterContentInit = function () {
        var element = this._element.nativeElement;
        if (this.selectableValue === undefined) {
            this.selectableValue = this._getIndex(element);
        }
        if (!!this.table) {
            this._bindListener();
        }
    };
    VdlDataTableSelectableRow = __decorate([
        Component({
            selector: 'tr[vdl-data-table-selectable-row]',
            template: "\n        <td class=\"vdl-data-check-cell\">\n          <vdl-checkbox [checked]=\"isActive\"></vdl-checkbox>\n        </td>\n        <ng-content></ng-content>\n    ",
            host: {
                '[class.active]': 'isActive',
                '(click)': 'change($event)'
            }
        }),
        __param(0, Optional()),
        __param(0, Inject(forwardRef(function () { return VdlDataTable; }))), 
        __metadata('design:paramtypes', [VdlDataTable, ElementRef])
    ], VdlDataTableSelectableRow);
    return VdlDataTableSelectableRow;
}(AbstractVdlDataTableSelectableRow));
