// vdl-angular/core/shared/hint/hint.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Input, Component, NgModule, ViewEncapsulation } from '@angular/core';
import { PlatformModule } from '../../platform/index';
import { CommonModule } from '@angular/common';
export var VdlHint = (function () {
    function VdlHint() {
    }
    VdlHint.prototype._isError = function () {
        if (this.error == '') {
            return true;
        }
    };
    __decorate([
        Input(), 
        __metadata('design:type', String)
    ], VdlHint.prototype, "error", void 0);
    __decorate([
        Input(), 
        __metadata('design:type', String)
    ], VdlHint.prototype, "hintText", void 0);
    VdlHint = __decorate([
        Component({selector: 'vdl-hint',
            host: {
                'class': 'vdl-hint',
                '[class.vdl-error]': '_isError()',
            },
            styles: [".vdl-hint{color:#414142}.vdl-hint.vdl-error{color:#b1181e}.vdl-hint~.vdl-hint::before{content:'-';margin:0 5px}"],
            template: "{{hintText}}",
            encapsulation: ViewEncapsulation.None,
        }), 
        __metadata('design:paramtypes', [])
    ], VdlHint);
    return VdlHint;
}());
export var VdlHintModule = (function () {
    function VdlHintModule() {
    }
    VdlHintModule.forRoot = function () {
        return {
            ngModule: VdlHintModule,
            providers: PlatformModule.forRoot().providers,
        };
    };
    VdlHintModule = __decorate([
        NgModule({
            declarations: [
                VdlHint
            ],
            imports: [
                CommonModule,
                PlatformModule
            ],
            exports: [
                VdlHint
            ],
        }), 
        __metadata('design:paramtypes', [])
    ], VdlHintModule);
    return VdlHintModule;
}());
