// vdl-angular/menu/menu.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { OverlayModule } from '../core';
import { VdlMenu } from './menu-directive';
import { VdlMenuItem } from './menu-item';
import { VdlMenuTrigger } from './menu-trigger';
import { VdlRippleModule } from '../core/ripple/ripple';
export { VdlMenu } from './menu-directive';
export { VdlMenuItem } from './menu-item';
export { VdlMenuTrigger } from './menu-trigger';
export var VdlMenuModule = (function () {
    function VdlMenuModule() {
    }
    /** @deprecated */
    VdlMenuModule.forRoot = function () {
        return {
            ngModule: VdlMenuModule,
            providers: [],
        };
    };
    VdlMenuModule = __decorate([
        NgModule({
            imports: [OverlayModule, CommonModule, VdlRippleModule],
            exports: [VdlMenu, VdlMenuItem, VdlMenuTrigger],
            declarations: [VdlMenu, VdlMenuItem, VdlMenuTrigger],
        }), 
        __metadata('design:paramtypes', [])
    ], VdlMenuModule);
    return VdlMenuModule;
}());
