// vdl-angular/menu/menu-directive.js
// TODO(kara): prevent-close functionality
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
import { Attribute, Component, ContentChildren, EventEmitter, Input, Output, QueryList, TemplateRef, ViewChild, ViewEncapsulation } from '@angular/core';
import { VdlMenuInvalidPositionX, VdlMenuInvalidPositionY } from './menu-errors';
import { VdlMenuItem } from './menu-item';
import { FocusKeyManager } from '../core/a11y/focus-key-manager';
import { transformMenu, fadeInItems } from './menu-animations';
export var VdlMenu = (function () {
    function VdlMenu(posX, posY) {
        /** Config object to be passed into the menu's ngClass */
        this._classList = {};
        /** Position of the menu in the X axis. */
        this.positionX = 'after';
        /** Position of the menu in the Y axis. */
        this.positionY = 'below';
        this.overlapTrigger = true;
        /** Event emitted when the menu is closed. */
        this.close = new EventEmitter();
        if (posX) {
            this._setPositionX(posX);
        }
        if (posY) {
            this._setPositionY(posY);
        }
        this.setPositionClasses(this.positionX, this.positionY);
    }
    VdlMenu.prototype.ngAfterContentInit = function () {
        var _this = this;
        this._keyManager = new FocusKeyManager(this.items).withWrap();
        this._tabSubscription = this._keyManager.tabOut.subscribe(function () {
            _this._emitCloseEvent();
        });
    };
    VdlMenu.prototype.ngOnDestroy = function () {
        this._tabSubscription.unsubscribe();
    };
    Object.defineProperty(VdlMenu.prototype, "classList", {
        /**
         * This method takes classes set on the host vdl-menu element and applies them on the
         * menu template that displays in the overlay container.  Otherwise, it's difficult
         * to style the containing menu from outside the component.
         * @param classes list of class names
         */
        set: function (classes) {
            this._classList = classes.split(' ').reduce(function (obj, className) {
                obj[className] = true;
                return obj;
            }, {});
            this.setPositionClasses(this.positionX, this.positionY);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Focus the first item in the menu. This method is used by the menu trigger
     * to focus the first item when the menu is opened by the ENTER key.
     */
    VdlMenu.prototype.focusFirstItem = function () {
        this._keyManager.setFirstItemActive();
    };
    /**
     * This emits a close event to which the trigger is subscribed. When emitted, the
     * trigger will close the menu.
     */
    VdlMenu.prototype._emitCloseEvent = function () {
        this.close.emit();
    };
    VdlMenu.prototype._setPositionX = function (pos) {
        if (pos !== 'before' && pos !== 'after') {
            throw new VdlMenuInvalidPositionX();
        }
        this.positionX = pos;
    };
    VdlMenu.prototype._setPositionY = function (pos) {
        if (pos !== 'above' && pos !== 'below') {
            throw new VdlMenuInvalidPositionY();
        }
        this.positionY = pos;
    };
    /**
     * It's necessary to set position-based classes to ensure the menu panel animation
     * folds out from the correct direction.
     */
    VdlMenu.prototype.setPositionClasses = function (posX, posY) {
        this._classList['vdl-menu-before'] = posX == 'before';
        this._classList['vdl-menu-after'] = posX == 'after';
        this._classList['vdl-menu-above'] = posY == 'above';
        this._classList['vdl-menu-below'] = posY == 'below';
    };
    __decorate([
        ViewChild(TemplateRef), 
        __metadata('design:type', TemplateRef)
    ], VdlMenu.prototype, "templateRef", void 0);
    __decorate([
        ContentChildren(VdlMenuItem), 
        __metadata('design:type', QueryList)
    ], VdlMenu.prototype, "items", void 0);
    __decorate([
        Input(), 
        __metadata('design:type', Object)
    ], VdlMenu.prototype, "overlapTrigger", void 0);
    __decorate([
        Input('class'), 
        __metadata('design:type', String), 
        __metadata('design:paramtypes', [String])
    ], VdlMenu.prototype, "classList", null);
    __decorate([
        Output(), 
        __metadata('design:type', Object)
    ], VdlMenu.prototype, "close", void 0);
    VdlMenu = __decorate([
        Component({selector: 'vdl-menu',
            host: { 'role': 'menu' },
            template: "<template><div class=\"vdl-menu-panel\" [ngClass]=\"_classList\" (keydown)=\"_keyManager.onKeydown($event)\" (click)=\"_emitCloseEvent()\" [@transformMenu]=\"'showing'\"><div class=\"vdl-menu-content\" [@fadeInItems]=\"'showing'\"><ng-content></ng-content></div></div></template>",
            styles: [".vdl-menu-panel{box-shadow:0 5px 5px -3px rgba(0,0,0,.2),0 8px 10px 1px rgba(0,0,0,.14),0 3px 14px 2px rgba(0,0,0,.12);min-width:112px;max-width:280px;overflow:auto;-webkit-overflow-scrolling:touch;max-height:calc(100vh + 48px)}.vdl-menu-panel.vdl-menu-after.vdl-menu-below{transform-origin:left top}.vdl-menu-panel.vdl-menu-after.vdl-menu-above{transform-origin:left bottom}.vdl-menu-panel.vdl-menu-before.vdl-menu-below{transform-origin:right top}.vdl-menu-panel.vdl-menu-before.vdl-menu-above{transform-origin:right bottom}[dir=rtl] .vdl-menu-panel.vdl-menu-after.vdl-menu-below{transform-origin:right top}[dir=rtl] .vdl-menu-panel.vdl-menu-after.vdl-menu-above{transform-origin:right bottom}[dir=rtl] .vdl-menu-panel.vdl-menu-before.vdl-menu-below{transform-origin:left top}[dir=rtl] .vdl-menu-panel.vdl-menu-before.vdl-menu-above{transform-origin:left bottom}@media screen and (-ms-high-contrast:active){.vdl-menu-panel{outline:solid 1px}}.vdl-menu-content{padding-top:8px;padding-bottom:8px}.vdl-menu-item{cursor:pointer;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;outline:0;border:none;white-space:nowrap;overflow-x:hidden;text-overflow:ellipsis;display:block;line-height:48px;height:48px;padding:0 16px;font-size:16px;font-family:Roboto,sans-serif;text-align:start;text-decoration:none;position:relative}.vdl-menu-item[disabled]{cursor:default}.vdl-menu-item .vdl-icon{margin-right:16px}[dir=rtl] .vdl-menu-item .vdl-icon{margin-left:16px}button.vdl-menu-item{width:100%}.vdl-menu-ripple{position:absolute;top:0;left:0;bottom:0;right:0}"],
            encapsulation: ViewEncapsulation.None,
            animations: [
                transformMenu,
                fadeInItems
            ],
            exportAs: 'vdlMenu'
        }),
        __param(0, Attribute('x-position')),
        __param(1, Attribute('y-position')), 
        __metadata('design:paramtypes', [String, String])
    ], VdlMenu);
    return VdlMenu;
}());
