// vdl-angular/input/index.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { NgModule } from '@angular/core';
import { VdlInputContainer, VdlInputDirective, VdlCharCount } from './input-container';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { PlatformModule } from '../core/platform/index';
import { VdlTextareaAutosize } from './autosize';
import { VdlIconModule } from '../icon/icon';
import { VdlPopoverModule } from '../popover/index';
import { VdlHintModule } from '../core/shared/hint/hint';
import { VdlButtonModule } from '../button/button';
export * from './autosize';
export * from './input-container';
export * from './input-container-errors';
export var VdlInputModule = (function () {
    function VdlInputModule() {
    }
    /** @deprecated */
    VdlInputModule.forRoot = function () {
        return {
            ngModule: VdlInputModule,
            providers: [],
        };
    };
    VdlInputModule = __decorate([
        NgModule({
            declarations: [
                VdlInputContainer,
                VdlCharCount,
                VdlTextareaAutosize,
                VdlInputDirective
            ],
            imports: [
                CommonModule,
                FormsModule,
                PlatformModule,
                VdlIconModule,
                VdlPopoverModule,
                VdlHintModule,
                VdlButtonModule
            ],
            exports: [
                VdlInputContainer,
                VdlCharCount,
                VdlTextareaAutosize,
                VdlInputDirective
            ],
        }), 
        __metadata('design:paramtypes', [])
    ], VdlInputModule);
    return VdlInputModule;
}());
