// vdl-angular/core/ripple/ripple.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { NgModule, Directive, ElementRef, Input, NgZone } from '@angular/core';
import { RippleRenderer } from './ripple-renderer';
import { ViewportRuler, VIEWPORT_RULER_PROVIDER } from '../overlay/position/viewport-ruler';
import { SCROLL_DISPATCHER_PROVIDER } from '../overlay/scroll/scroll-dispatcher';
export var VdlRipple = (function () {
    function VdlRipple(elementRef, ngZone, ruler) {
        /**
         * If set, the radius in pixels of foreground ripples when fully expanded. If unset, the radius
         * will be the distance from the center of the ripple to the furthest corner of the host element's
         * bounding rectangle.
         */
        this.radius = 0;
        /**
         * If set, the normal duration of ripple animations is divided by this value. For example,
         * setting it to 0.5 will cause the animations to take twice as long.
         * A changed speedFactor will not modify the fade-out duration of the ripples.
         */
        this.speedFactor = 1;
        this._rippleRenderer = new RippleRenderer(elementRef, ngZone, ruler);
    }
    VdlRipple.prototype.ngOnChanges = function (changes) {
        if (changes['trigger'] && this.trigger) {
            this._rippleRenderer.setTriggerElement(this.trigger);
        }
        this._rippleRenderer.rippleDisabled = this.disabled;
        this._updateRippleConfig();
    };
    VdlRipple.prototype.ngOnDestroy = function () {
        // Set the trigger element to null to cleanup all listeners.
        this._rippleRenderer.setTriggerElement(null);
    };
    /** Launches a manual ripple at the specified position. */
    VdlRipple.prototype.launch = function (pageX, pageY, config) {
        this._rippleRenderer.fadeInRipple(pageX, pageY, config);
    };
    /** Updates the ripple configuration with the input values. */
    VdlRipple.prototype._updateRippleConfig = function () {
        this._rippleRenderer.rippleConfig = {
            centered: this.centered,
            speedFactor: this.speedFactor,
            radius: this.radius,
            color: this.color
        };
    };
    __decorate([
        Input('vdlRippleTrigger'), 
        __metadata('design:type', Object)
    ], VdlRipple.prototype, "trigger", void 0);
    __decorate([
        Input('vdlRippleCentered'), 
        __metadata('design:type', Boolean)
    ], VdlRipple.prototype, "centered", void 0);
    __decorate([
        Input('vdlRippleDisabled'), 
        __metadata('design:type', Boolean)
    ], VdlRipple.prototype, "disabled", void 0);
    __decorate([
        Input('vdlRippleRadius'), 
        __metadata('design:type', Number)
    ], VdlRipple.prototype, "radius", void 0);
    __decorate([
        Input('vdlRippleSpeedFactor'), 
        __metadata('design:type', Number)
    ], VdlRipple.prototype, "speedFactor", void 0);
    __decorate([
        Input('vdlRippleColor'), 
        __metadata('design:type', String)
    ], VdlRipple.prototype, "color", void 0);
    __decorate([
        Input('vdlRippleUnbounded'), 
        __metadata('design:type', Boolean)
    ], VdlRipple.prototype, "unbounded", void 0);
    VdlRipple = __decorate([
        Directive({
            selector: '[vdl-ripple]',
            host: {
                '[class.vdl-ripple]': 'true',
                '[class.vdl-ripple-unbounded]': 'unbounded'
            }
        }), 
        __metadata('design:paramtypes', [ElementRef, NgZone, ViewportRuler])
    ], VdlRipple);
    return VdlRipple;
}());
export var VdlRippleModule = (function () {
    function VdlRippleModule() {
    }
    /** @deprecated */
    VdlRippleModule.forRoot = function () {
        return {
            ngModule: VdlRippleModule,
            providers: []
        };
    };
    VdlRippleModule = __decorate([
        NgModule({
            exports: [VdlRipple],
            declarations: [VdlRipple],
            providers: [VIEWPORT_RULER_PROVIDER, SCROLL_DISPATCHER_PROVIDER],
        }), 
        __metadata('design:paramtypes', [])
    ], VdlRippleModule);
    return VdlRippleModule;
}());
