// vdl-angular/card/card.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { NgModule, Component, ViewEncapsulation, ChangeDetectionStrategy, Directive } from '@angular/core';
/**
 * Content of a card, needed as it's used as a selector in the API.
 */
export var VdlCardContent = (function () {
    function VdlCardContent() {
    }
    VdlCardContent = __decorate([
        Directive({
            selector: 'vdl-card-content',
            host: {
                '[class.vdl-card-content]': 'true'
            }
        }), 
        __metadata('design:paramtypes', [])
    ], VdlCardContent);
    return VdlCardContent;
}());
/**
 * Title of a card, needed as it's used as a selector in the API.
 */
export var VdlCardTitle = (function () {
    function VdlCardTitle() {
    }
    VdlCardTitle = __decorate([
        Directive({
            selector: 'vdl-card-title',
            host: {
                '[class.vdl-card-title]': 'true'
            }
        }), 
        __metadata('design:paramtypes', [])
    ], VdlCardTitle);
    return VdlCardTitle;
}());
/**
 * Sub-title of a card, needed as it's used as a selector in the API.
 */
export var VdlCardSubtitle = (function () {
    function VdlCardSubtitle() {
    }
    VdlCardSubtitle = __decorate([
        Directive({
            selector: 'vdl-card-subtitle',
            host: {
                '[class.vdl-card-subtitle]': 'true'
            }
        }), 
        __metadata('design:paramtypes', [])
    ], VdlCardSubtitle);
    return VdlCardSubtitle;
}());
/**
 * Header Text section of a card, needed as it's used as a selector in the API.
 */
export var VdlCardHeaderText = (function () {
    function VdlCardHeaderText() {
    }
    VdlCardHeaderText = __decorate([
        Directive({
            selector: 'vdl-card-header-text',
            host: {
                '[class.vdl-card-header-text]': 'true'
            }
        }), 
        __metadata('design:paramtypes', [])
    ], VdlCardHeaderText);
    return VdlCardHeaderText;
}());
/**
 * Header Content section of a card, needed as it's used as a selector in the API.
 */
export var VdlCardHeaderContent = (function () {
    function VdlCardHeaderContent() {
    }
    VdlCardHeaderContent = __decorate([
        Directive({
            selector: 'vdl-card-header-content',
            host: {
                '[class.vdl-card-header-content]': 'true'
            }
        }), 
        __metadata('design:paramtypes', [])
    ], VdlCardHeaderContent);
    return VdlCardHeaderContent;
}());
/**
 * Header Action section of a card, needed as it's used as a selector in the API.
 */
export var VdlCardHeaderActions = (function () {
    function VdlCardHeaderActions() {
    }
    VdlCardHeaderActions = __decorate([
        Directive({
            selector: 'vdl-card-header-actions',
            host: {
                '[class.vdl-card-header-actions]': 'true'
            }
        }), 
        __metadata('design:paramtypes', [])
    ], VdlCardHeaderActions);
    return VdlCardHeaderActions;
}());
/**
 * Action section of a card, needed as it's used as a selector in the API.
 */
export var VdlCardActions = (function () {
    function VdlCardActions() {
    }
    VdlCardActions = __decorate([
        Directive({
            selector: 'vdl-card-actions',
            host: {
                '[class.vdl-card-actions]': 'true'
            }
        }), 
        __metadata('design:paramtypes', [])
    ], VdlCardActions);
    return VdlCardActions;
}());
/**
 * Footer of a card, needed as it's used as a selector in the API.
 */
export var VdlCardFooter = (function () {
    function VdlCardFooter() {
    }
    VdlCardFooter = __decorate([
        Directive({
            selector: 'vdl-card-footer',
            host: {
                '[class.vdl-card-footer]': 'true'
            }
        }), 
        __metadata('design:paramtypes', [])
    ], VdlCardFooter);
    return VdlCardFooter;
}());
/**
 * Image used in a card, needed to add the vdl- CSS styling.
 */
export var VdlCardSmImage = (function () {
    function VdlCardSmImage() {
    }
    VdlCardSmImage = __decorate([
        Directive({
            selector: '[vdl-card-sm-image]',
            host: {
                '[class.vdl-card-sm-image]': 'true'
            }
        }), 
        __metadata('design:paramtypes', [])
    ], VdlCardSmImage);
    return VdlCardSmImage;
}());
/**
 * Image used in a card, needed to add the vdl- CSS styling.
 */
export var VdlCardVdlImage = (function () {
    function VdlCardVdlImage() {
    }
    VdlCardVdlImage = __decorate([
        Directive({
            selector: '[vdl-card-vdl-image]',
            host: {
                '[class.vdl-card-vdl-image]': 'true'
            }
        }), 
        __metadata('design:paramtypes', [])
    ], VdlCardVdlImage);
    return VdlCardVdlImage;
}());
/**
 * Image used in a card, needed to add the vdl- CSS styling.
 */
export var VdlCardLgImage = (function () {
    function VdlCardLgImage() {
    }
    VdlCardLgImage = __decorate([
        Directive({
            selector: '[vdl-card-lg-image]',
            host: {
                'class.vdl-card-lg-image': 'true'
            }
        }), 
        __metadata('design:paramtypes', [])
    ], VdlCardLgImage);
    return VdlCardLgImage;
}());
/**
 * Image used in a card, needed to add the vdl- CSS styling.
 */
export var VdlCardImage = (function () {
    function VdlCardImage() {
    }
    VdlCardImage = __decorate([
        Directive({
            selector: '[vdl-card-image]',
            host: {
                '[class.vdl-card-image]': 'true'
            }
        }), 
        __metadata('design:paramtypes', [])
    ], VdlCardImage);
    return VdlCardImage;
}());
/**
 * Large image used in a card, needed to add the vdl- CSS styling.
 */
export var VdlCardXlImage = (function () {
    function VdlCardXlImage() {
    }
    VdlCardXlImage = __decorate([
        Directive({
            selector: 'vdl-card-xl-image',
            host: {
                '[class.vdl-card-xl-image]': 'true'
            }
        }), 
        __metadata('design:paramtypes', [])
    ], VdlCardXlImage);
    return VdlCardXlImage;
}());
/**
 * Avatar image used in a card, needed to add the vdl- CSS styling.
 */
export var VdlCardAvatar = (function () {
    function VdlCardAvatar() {
    }
    VdlCardAvatar = __decorate([
        Directive({
            selector: 'vdl-card-avatar',
            host: {
                '[class.vdl-card-avatar]': 'true'
            }
        }), 
        __metadata('design:paramtypes', [])
    ], VdlCardAvatar);
    return VdlCardAvatar;
}());
/**
 * A basic content container component that adds the styles of a Material design card.
 *
 * While this component can be used alone, it also provides a number
 * of preset styles for common card sections, including:
 * - vdl-card-title
 * - vdl-card-subtitle
 * - vdl-card-content
 * - vdl-card-actions
 * - vdl-card-footer
 */
export var VdlCard = (function () {
    function VdlCard() {
    }
    VdlCard = __decorate([
        Component({selector: 'vdl-card',
            template: "<ng-content></ng-content>",
            styles: [".vdl-card{border:1px solid #e5e5e5;border-bottom-width:2px;display:block;position:relative;padding:24px;border-radius:2px;font-family:Roboto,sans-serif;background-color:#fff;color:#414142}@media screen and (-ms-high-contrast:active){.vdl-card{outline:solid 1px}}.vdl-card-flat{box-shadow:none}.vdl-card-actions,.vdl-card-content,.vdl-card-subtitle,.vdl-card-title{display:block;margin-bottom:16px}.vdl-card-title{font-size:1.5rem;font-weight:400}.vdl-card-subtitle{font-size:14px}.vdl-card-content{font-size:14px}.vdl-card-actions{margin-left:-16px;margin-right:-16px;padding:8px 0}.vdl-card-actions[align=end]{display:flex;justify-content:flex-end}.vdl-card-image{width:calc(100% + 48px);margin:0 -24px 16px -24px}.vdl-card-xl-image{width:240px;height:240px;margin:-8px}.vdl-card-footer{position:absolute;width:100%;min-height:5px;bottom:0;left:0}.vdl-card-actions .vdl-button,.vdl-card-actions .vdl-raised-button{margin:0 4px}.vdl-card-header{flex:1 1 auto;display:flex;flex-direction:row;height:auto;margin:-8px 0 16px 0}.vdl-card-header-text{height:auto;margin:0 8px;flex:1 1 auto}.vdl-card-header-content{flex:0 1 auto}.vdl-card-header-image{flex:0 1 auto}.vdl-card-avatar{height:auto;width:auto}.vdl-card-header-actions{flex:0 1 auto;background-color:#fff}.vdl-card-lg-image,.vdl-card-sm-image,.vdl-card-vdl-image{margin:-8px 0}.vdl-card-title-group{display:flex;justify-content:space-between;margin:0 -8px}.vdl-card-sm-image{width:80px;height:80px}.vdl-card-vdl-image{width:112px;height:112px}.vdl-card-lg-image{width:152px;height:152px}@media (max-width:600px){.vdl-card{padding:24px 16px}.vdl-card-actions{margin-left:-8px;margin-right:-8px}.vdl-card-image{width:calc(100% + 32px);margin:16px -16px}.vdl-card-title-group{margin:0}.vdl-card-xl-image{margin-left:0;margin-right:0}.vdl-card-header{margin:-8px 0 0 0}}.vdl-card-content>:first-child,.vdl-card>:first-child{margin-top:0}.vdl-card-content>:last-child,.vdl-card>:last-child{margin-bottom:0}.vdl-card-image:first-child{margin-top:-24px}.vdl-card>.vdl-card-actions:last-child{margin-bottom:-16px;padding-bottom:0}.vdl-card-actions .vdl-button:first-child,.vdl-card-actions .vdl-raised-button:first-child{margin-left:0;margin-right:0}.vdl-card-subtitle:not(:first-child),.vdl-card-title:not(:first-child){margin-top:-4px}.vdl-card-header .vdl-card-subtitle:not(:first-child){margin-top:-8px}.vdl-card>.vdl-card-xl-image:first-child{margin-top:-8px}.vdl-card>.vdl-card-xl-image:last-child{margin-bottom:-8px}"],
            encapsulation: ViewEncapsulation.None,
            changeDetection: ChangeDetectionStrategy.OnPush,
            host: {
                '[class.vdl-card]': 'true'
            }
        }), 
        __metadata('design:paramtypes', [])
    ], VdlCard);
    return VdlCard;
}());
/**
 * Component intended to be used within the `<vdl-card>` component. It adds styles for a
 * preset header section (i.e. a title, subtitle, and avatar layout).
 */
export var VdlCardHeader = (function () {
    function VdlCardHeader() {
    }
    VdlCardHeader = __decorate([
        Component({selector: 'vdl-card-header',
            template: "<ng-content select=\"[vdl-card-avatar], [vdl-card-header-image]\"></ng-content><vdl-card-header-text class=\"vdl-card-header-text\"><ng-content select=\"vdl-card-title, vdl-card-subtitle\"></ng-content></vdl-card-header-text><vdl-card-header-content class=\"vdl-card-header-content\"><ng-content></ng-content></vdl-card-header-content><ng-content select=\"vdl-card-header-actions\"></ng-content>",
            encapsulation: ViewEncapsulation.None,
            changeDetection: ChangeDetectionStrategy.OnPush,
            host: {
                '[class.vdl-card-header]': 'true'
            }
        }), 
        __metadata('design:paramtypes', [])
    ], VdlCardHeader);
    return VdlCardHeader;
}());
/**
 * Component intended to be used within the <vdl-card> component. It adds styles for a preset
 * layout that groups an image with a title section.
 */
export var VdlCardTitleGroup = (function () {
    function VdlCardTitleGroup() {
    }
    VdlCardTitleGroup = __decorate([
        Component({selector: 'vdl-card-title-group',
            template: "<div><ng-content select=\"vdl-card-title, vdl-card-subtitle\"></ng-content></div><ng-content select=\"img\"></ng-content><ng-content></ng-content>",
            encapsulation: ViewEncapsulation.None,
            changeDetection: ChangeDetectionStrategy.OnPush,
            host: {
                '[class.vdl-card-title-group]': 'true'
            }
        }), 
        __metadata('design:paramtypes', [])
    ], VdlCardTitleGroup);
    return VdlCardTitleGroup;
}());
export var VdlCardModule = (function () {
    function VdlCardModule() {
    }
    /** @deprecated */
    VdlCardModule.forRoot = function () {
        return {
            ngModule: VdlCardModule,
            providers: []
        };
    };
    VdlCardModule = __decorate([
        NgModule({
            exports: [
                VdlCard,
                VdlCardHeader,
                VdlCardTitleGroup,
                VdlCardContent,
                VdlCardTitle,
                VdlCardSubtitle,
                VdlCardHeaderText,
                VdlCardHeaderContent,
                VdlCardHeaderActions,
                VdlCardActions,
                VdlCardFooter,
                VdlCardSmImage,
                VdlCardVdlImage,
                VdlCardLgImage,
                VdlCardImage,
                VdlCardXlImage,
                VdlCardAvatar,
            ],
            declarations: [
                VdlCard, VdlCardHeader, VdlCardTitleGroup, VdlCardContent, VdlCardTitle, VdlCardSubtitle,
                VdlCardActions, VdlCardFooter, VdlCardSmImage, VdlCardVdlImage, VdlCardLgImage, VdlCardImage,
                VdlCardXlImage, VdlCardAvatar, VdlCardHeaderText, VdlCardHeaderContent, VdlCardHeaderActions
            ],
        }), 
        __metadata('design:paramtypes', [])
    ], VdlCardModule);
    return VdlCardModule;
}());
