// vdl-angular/popover/popover-predefined.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component, ChangeDetectionStrategy, ViewEncapsulation, Input } from '@angular/core';
import { VdlIconRegistry } from '../icon/icon-registry';
export var VdlPopoverPredefined = (function () {
    function VdlPopoverPredefined(vdlIconRegistry) {
        this.popoverPlacement = 'auto';
        vdlIconRegistry.registerFontClassAlias('fontawesome', 'fa');
    }
    __decorate([
        Input(), 
        __metadata('design:type', String)
    ], VdlPopoverPredefined.prototype, "popoverTitle", void 0);
    __decorate([
        Input(), 
        __metadata('design:type', Object)
    ], VdlPopoverPredefined.prototype, "popoverPlacement", void 0);
    __decorate([
        Input(), 
        __metadata('design:type', Object)
    ], VdlPopoverPredefined.prototype, "popoverType", void 0);
    __decorate([
        Input(), 
        __metadata('design:type', String)
    ], VdlPopoverPredefined.prototype, "popoverIcon", void 0);
    VdlPopoverPredefined = __decorate([
        Component({selector: 'vdl-popover-predefined',
            changeDetection: ChangeDetectionStrategy.OnPush,
            encapsulation: ViewEncapsulation.None,
            template: "<template #popContentInfo><ng-content></ng-content></template><button vdl-icon-button [vdlPopover]=\"popContentInfo\" [popoverTitle]=\"popoverTitle\" [placement]=\"popoverPlacement\"><vdl-icon *ngIf=\"popoverType == 'info'\" class=\"vdl-info-circle-icon\" fontSet=\"fontawesome\" fontIcon=\"fa-info-circle\"></vdl-icon><vdl-icon *ngIf=\"popoverType == 'error'\" class=\"vdl-error-circle-icon\" fontSet=\"fontawesome\" fontIcon=\"fa-exclamation-circle\"></vdl-icon><vdl-icon *ngIf=\"popoverType == 'custom'\" class=\"vdl-custom-icon\" fontSet=\"fontawesome\" fontIcon=\"{{popoverIcon}}\"></vdl-icon></button>",
            styles: ["vdl-popover-predefined button:focus vdl-icon{color:#005a99}vdl-popover-predefined button[vdl-tertiary-button]{padding:0;background-color:transparent;border-color:transparent;line-height:0;color:#000;cursor:default}vdl-popover-predefined button[vdl-tertiary-button].hover,vdl-popover-predefined button[vdl-tertiary-button]:hover{color:#000;background-color:transparent;border-color:transparent}vdl-popover-predefined button[vdl-tertiary-button].focus,vdl-popover-predefined button[vdl-tertiary-button]:focus{color:#000;background-color:transparent;border-color:transparent}vdl-popover-predefined button[vdl-tertiary-button].active,vdl-popover-predefined button[vdl-tertiary-button]:active{color:#000;background-color:transparent;border-color:transparent}vdl-popover-predefined .vdl-info-circle-icon{color:#414142;cursor:pointer}vdl-popover-predefined .vdl-error-circle-icon{color:#b1181e;cursor:pointer}vdl-popover-predefined .vdl-custom-icon{color:#414142;cursor:pointer}"],
            viewProviders: [VdlIconRegistry],
        }), 
        __metadata('design:paramtypes', [VdlIconRegistry])
    ], VdlPopoverPredefined);
    return VdlPopoverPredefined;
}());
