// vdl-angular/top-bar/index.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VdlTopBar, VdlTopBarLogo, VdlTopBarTitle, VdlTopBarActions } from './top-bar';
import { VdlInputModule } from '../input/index';
import { VdlIconModule } from '../icon/icon';
import { VdlButtonModule } from '../button/button';
export * from './top-bar';
export var VdlTopBarModule = (function () {
    function VdlTopBarModule() {
    }
    VdlTopBarModule.forRoot = function () {
        return {
            ngModule: VdlTopBarModule,
            providers: []
        };
    };
    VdlTopBarModule = __decorate([
        NgModule({
            imports: [
                CommonModule,
                VdlInputModule,
                VdlIconModule,
                VdlButtonModule
            ],
            exports: [
                VdlTopBar,
                VdlTopBarTitle,
                VdlTopBarLogo,
                VdlTopBarActions,
            ],
            declarations: [
                VdlTopBar,
                VdlTopBarTitle,
                VdlTopBarLogo,
                VdlTopBarActions
            ],
        }), 
        __metadata('design:paramtypes', [])
    ], VdlTopBarModule);
    return VdlTopBarModule;
}());
