// vdl-angular/top-bar/top-bar.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component, Directive, ViewEncapsulation, ChangeDetectionStrategy } from '@angular/core';
import { VdlIconRegistry } from '../icon/icon-registry';
export var VdlTopBarLogo = (function () {
    function VdlTopBarLogo() {
    }
    VdlTopBarLogo = __decorate([
        Directive({
            selector: '[vdl-top-bar-logo]',
            host: {
                '[class.vdl-top-bar-logo]': 'true'
            }
        }), 
        __metadata('design:paramtypes', [])
    ], VdlTopBarLogo);
    return VdlTopBarLogo;
}());
export var VdlTopBarTitle = (function () {
    function VdlTopBarTitle() {
    }
    VdlTopBarTitle = __decorate([
        Directive({
            selector: '[vdl-top-bar-title]',
            host: {
                '[class.vdl-top-bar-title]': 'true'
            }
        }), 
        __metadata('design:paramtypes', [])
    ], VdlTopBarTitle);
    return VdlTopBarTitle;
}());
export var VdlTopBarActions = (function () {
    function VdlTopBarActions() {
    }
    VdlTopBarActions = __decorate([
        Directive({
            selector: 'vdl-top-bar-actions',
            host: {
                '[class.vdl-top-bar-actions]': 'true'
            }
        }), 
        __metadata('design:paramtypes', [])
    ], VdlTopBarActions);
    return VdlTopBarActions;
}());
export var VdlTopBar = (function () {
    function VdlTopBar() {
    }
    VdlTopBar = __decorate([
        Component({selector: 'vdl-top-bar',
            template: "<div><ng-content select=\"[vdl-top-bar-logo]\"></ng-content></div><ng-content select=\"[vdl-top-bar-title]\"></ng-content><ng-content></ng-content><ng-content select=\"vdl-top-bar-actions\"></ng-content>",
            styles: [".vdl-top-bar{height:59px;background-color:#f9f9f9;border-bottom:solid 1px #c2c2c2;flex-direction:row;display:flex;flex-wrap:nowrap;justify-content:space-between}.vdl-top-bar .vdl-top-bar-logo{width:40px;height:60px;padding-left:10px}.vdl-top-bar .vdl-top-bar-title{font-family:Roboto,sans-serif;color:#414142;font-size:22px;padding:20px 30px 20px 10px;line-height:22px;min-width:220px;flex:1 1 auto;margin-right:15px}.vdl-top-bar vdl-input-container{margin-top:3px;flex:1 1 auto;margin-left:15px;margin-right:20px}.vdl-top-bar vdl-input-container .vdl-input-element{margin-top:0;margin-bottom:8px}.vdl-top-bar vdl-input-container .vdl-input-underline{margin-top:11px}.vdl-top-bar vdl-input-container .vdl-input-underline.vdl-focused{border-top-width:2px;margin-top:10px}.vdl-top-bar .vdl-top-bar-actions{display:flex;align-self:center;border-left:dotted 1px #000}.vdl-top-bar .vdl-top-bar-actions>button{padding:0 20px;margin:0 20px;flex:0 1 auto;-ms-grid-row-align:center;align-self:center}.vdl-top-bar button[vdl-icon-button]{text-align:center;font-size:20px}.vdl-top-bar button[vdl-icon-button] .vdl-search{padding-right:20px}.vdl-top-bar button[vdl-icon-button] .vdl-search-top-bar{padding-right:40px}.vdl-top-bar button[vdl-icon-button] vdl-icon{cursor:pointer;height:20px;width:20px;font-size:20px;line-height:20px}"],
            host: {
                '[class.vdl-top-bar]': 'true'
            },
            viewProviders: [VdlIconRegistry],
            encapsulation: ViewEncapsulation.None,
            changeDetection: ChangeDetectionStrategy.OnPush
        }), 
        __metadata('design:paramtypes', [])
    ], VdlTopBar);
    return VdlTopBar;
}());
