// vdl-angular/dialog/dialog-content-directives.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Directive, Input } from '@angular/core';
import { VdlDialogRef } from './dialog-ref';
/**
 * Button that will close the current dialog.
 */
export var VdlDialogClose = (function () {
    function VdlDialogClose(dialogRef) {
        this.dialogRef = dialogRef;
        /** Screenreader label for the button. */
        this.ariaLabel = 'Close dialog';
    }
    __decorate([
        Input('aria-label'), 
        __metadata('design:type', String)
    ], VdlDialogClose.prototype, "ariaLabel", void 0);
    VdlDialogClose = __decorate([
        Directive({
            selector: 'button[vdl-dialog-close]',
            host: {
                '(click)': 'dialogRef.close()',
                '[attr.aria-label]': 'ariaLabel',
                'type': 'button',
            }
        }), 
        __metadata('design:paramtypes', [VdlDialogRef])
    ], VdlDialogClose);
    return VdlDialogClose;
}());
/**
 * Title of a dialog element. Stays fixed to the top of the dialog when scrolling.
 */
export var VdlDialogTitle = (function () {
    function VdlDialogTitle() {
    }
    VdlDialogTitle = __decorate([
        Directive({
            selector: '[vdl-dialog-title]',
            host: {
                '[class.vdl-dialog-title]': 'true'
            }
        }), 
        __metadata('design:paramtypes', [])
    ], VdlDialogTitle);
    return VdlDialogTitle;
}());
/**
 * Scrollable content container of a dialog.
 */
export var VdlDialogContent = (function () {
    function VdlDialogContent() {
    }
    VdlDialogContent = __decorate([
        Directive({
            selector: '[vdl-dialog-content], vdl-dialog-content',
            host: {
                '[class.vdl-dialog-content]': 'true'
            }
        }), 
        __metadata('design:paramtypes', [])
    ], VdlDialogContent);
    return VdlDialogContent;
}());
/**
 * Container for the bottom action buttons in a dialog.
 * Stays fixed to the bottom when scrolling.
 */
export var VdlDialogActions = (function () {
    function VdlDialogActions() {
    }
    VdlDialogActions = __decorate([
        Directive({
            selector: '[vdl-dialog-actions], vdl-dialog-actions',
            host: {
                '[class.vdl-dialog-actions]': 'true'
            }
        }), 
        __metadata('design:paramtypes', [])
    ], VdlDialogActions);
    return VdlDialogActions;
}());
