// vdl-angular/dialog/index.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { NgModule } from '@angular/core';
import { OverlayModule, PortalModule, A11yModule } from '../core';
import { VdlDialog } from './dialog';
import { VdlDialogContainer } from './dialog-container';
import { VdlDialogClose, VdlDialogContent, VdlDialogTitle, VdlDialogActions } from './dialog-content-directives';
export var VdlDialogModule = (function () {
    function VdlDialogModule() {
    }
    /** @deprecated */
    VdlDialogModule.forRoot = function () {
        return {
            ngModule: VdlDialogModule,
            providers: [],
        };
    };
    VdlDialogModule = __decorate([
        NgModule({
            imports: [
                OverlayModule,
                PortalModule,
                A11yModule,
            ],
            exports: [
                VdlDialogContainer,
                VdlDialogClose,
                VdlDialogTitle,
                VdlDialogContent,
                VdlDialogActions,
            ],
            declarations: [
                VdlDialogContainer,
                VdlDialogClose,
                VdlDialogTitle,
                VdlDialogActions,
                VdlDialogContent,
            ],
            providers: [
                VdlDialog,
            ],
            entryComponents: [VdlDialogContainer],
        }), 
        __metadata('design:paramtypes', [])
    ], VdlDialogModule);
    return VdlDialogModule;
}());
export * from './dialog';
export * from './dialog-container';
export * from './dialog-content-directives';
export * from './dialog-config';
export * from './dialog-ref';
export { VDL_DIALOG_DATA } from './dialog-injector';
