// vdl-angular/tabs/tab-nav-bar/tab-nav-bar.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component, Input, ViewChild, ElementRef, ViewEncapsulation, Directive } from '@angular/core';
import { VdlInkBar } from '../ink-bar';
/**
 * Navigation component matching the styles of the tab group header.
 * Provides anchored navigation with animated ink bar.
 */
export var VdlTabNavBar = (function () {
    function VdlTabNavBar() {
    }
    /** Animates the ink bar to the position of the active link element. */
    VdlTabNavBar.prototype.updateActiveLink = function (element) {
        this._inkBar.alignToElement(element);
    };
    __decorate([
        ViewChild(VdlInkBar), 
        __metadata('design:type', VdlInkBar)
    ], VdlTabNavBar.prototype, "_inkBar", void 0);
    VdlTabNavBar = __decorate([
        Component({selector: '[vdl-tab-nav-bar]',
            template: "<ng-content></ng-content><vdl-ink-bar></vdl-ink-bar>",
            styles: ["[vdl-tab-nav-bar]{overflow:hidden;position:relative;display:flex;flex-direction:row;flex-shrink:0}[vdl-tab-link]{line-height:40px;height:40px;margin:0;padding:0 15px;font-size:16px;font-family:Roboto,sans-serif;font-weight:500;cursor:pointer;box-sizing:border-box;color:#414142;min-width:160px;text-align:center;border-radius:3px;transition:border-bottom 4ms ease-in-out 3ms;text-decoration:none;position:relative;overflow:hidden;border-radius:3px;margin:0}[vdl-tab-link].focus,[vdl-tab-link]:focus{outline:0}[vdl-tab-link].hover,[vdl-tab-link]:hover{background-color:#f3f3f3}[vdl-tab-link].vdl-tab-label-active,[vdl-tab-link].vdl-tab-link-active{color:#005a99}[vdl-tab-link].vdl-tab-disabled{cursor:default;pointer-events:none;color:#a5a5a5}[vdl-tab-link]:focus{outline-offset:-2px;outline-width:.5px;outline-style:auto;outline-color:#005a99}@media (max-width:600px){[vdl-tab-link]{min-width:72px}}vdl-ink-bar{position:absolute;bottom:0;height:2px;background-color:#005a99;transition:.5s cubic-bezier(.35,0,.25,1)}.vdl-tab-group-inverted-header vdl-ink-bar{bottom:auto;top:0}"],
            encapsulation: ViewEncapsulation.None,
        }), 
        __metadata('design:paramtypes', [])
    ], VdlTabNavBar);
    return VdlTabNavBar;
}());
/**
 * Link inside of a `vdl-tab-nav-bar`.
 */
export var VdlTabLink = (function () {
    function VdlTabLink(_vdlTabNavBar, _element) {
        this._vdlTabNavBar = _vdlTabNavBar;
        this._element = _element;
        this._isActive = false;
    }
    Object.defineProperty(VdlTabLink.prototype, "active", {
        /** Whether the link is active. */
        get: function () { return this._isActive; },
        set: function (value) {
            this._isActive = value;
            if (value) {
                // todo(dhm): triggers updating the ink-bar which we should probably remove in favor of simple border bottom
                this._vdlTabNavBar.updateActiveLink(this._element.nativeElement);
            }
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        Input(), 
        __metadata('design:type', Boolean)
    ], VdlTabLink.prototype, "active", null);
    VdlTabLink = __decorate([
        Directive({
            selector: '[vdl-tab-link]',
            host: {
                '[class.vdl-tab-link-active]': 'active === true'
            }
        }), 
        __metadata('design:paramtypes', [VdlTabNavBar, ElementRef])
    ], VdlTabLink);
    return VdlTabLink;
}());
