// vdl-angular/grid-list/grid-list.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
import { NgModule, Component, ViewEncapsulation, Input, ContentChildren, QueryList, Renderer, ElementRef, Optional } from '@angular/core';
import { VdlGridTile, VdlGridTileText, VdlGridTileFooterCssMatStyler, VdlGridTileHeaderCssMatStyler, VdlGridAvatarCssMatStyler } from './grid-tile';
import { TileCoordinator } from './tile-coordinator';
import { FitTileStyler, RatioTileStyler, FixedTileStyler } from './tile-styler';
import { VdlGridListColsError } from './grid-list-errors';
import { Dir, VdlLineModule } from '../core';
import { coerceToString, coerceToNumber } from './grid-list-measure';
// TODO(kara): Conditional (responsive) column count / row size.
// TODO(kara): Re-layout on window resize / media change (debounced).
// TODO(kara): gridTileHeader and gridTileFooter.
var VDL_FIT_MODE = 'fit';
export var VdlGridList = (function () {
    function VdlGridList(_renderer, _element, _dir) {
        this._renderer = _renderer;
        this._element = _element;
        this._dir = _dir;
        /** The amount of space between tiles. This will be something like '5px' or '2em'. */
        this._gutter = '1px';
    }
    Object.defineProperty(VdlGridList.prototype, "cols", {
        /** Amount of columns in the grid list. */
        get: function () { return this._cols; },
        set: function (value) { this._cols = coerceToNumber(value); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(VdlGridList.prototype, "gutterSize", {
        /** Size of the grid list's gutter in pixels. */
        get: function () { return this._gutter; },
        set: function (value) { this._gutter = coerceToString(value); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(VdlGridList.prototype, "rowHeight", {
        /** Set internal representation of row height from the user-provided value. */
        set: function (value) {
            this._rowHeight = coerceToString(value);
            this._setTileStyler();
        },
        enumerable: true,
        configurable: true
    });
    VdlGridList.prototype.ngOnInit = function () {
        this._checkCols();
        this._checkRowHeight();
    };
    /**
     * The layout calculation is fairly cheap if nothing changes, so there's little cost
     * to run it frequently.
     */
    VdlGridList.prototype.ngAfterContentChecked = function () {
        this._layoutTiles();
    };
    /** Throw a friendly error if cols property is missing */
    VdlGridList.prototype._checkCols = function () {
        if (!this.cols) {
            throw new VdlGridListColsError();
        }
    };
    /** Default to equal width:height if rowHeight property is missing */
    VdlGridList.prototype._checkRowHeight = function () {
        if (!this._rowHeight) {
            this._tileStyler = new RatioTileStyler('1:1');
        }
    };
    /** Creates correct Tile Styler subtype based on rowHeight passed in by user */
    VdlGridList.prototype._setTileStyler = function () {
        if (this._rowHeight === VDL_FIT_MODE) {
            this._tileStyler = new FitTileStyler();
        }
        else if (this._rowHeight && this._rowHeight.indexOf(':') > -1) {
            this._tileStyler = new RatioTileStyler(this._rowHeight);
        }
        else {
            this._tileStyler = new FixedTileStyler(this._rowHeight);
        }
    };
    /** Computes and applies the size and position for all children grid tiles. */
    VdlGridList.prototype._layoutTiles = function () {
        var _this = this;
        var tracker = new TileCoordinator(this.cols, this._tiles);
        var direction = this._dir ? this._dir.value : 'ltr';
        this._tileStyler.init(this.gutterSize, tracker, this.cols, direction);
        this._tiles.forEach(function (tile, index) {
            var pos = tracker.positions[index];
            _this._tileStyler.setStyle(tile, pos.row, pos.col);
        });
        this._setListStyle(this._tileStyler.getComputedHeight());
    };
    /** Sets style on the main grid-list element, given the style name and value. */
    VdlGridList.prototype._setListStyle = function (style) {
        if (style) {
            this._renderer.setElementStyle(this._element.nativeElement, style[0], style[1]);
        }
    };
    __decorate([
        ContentChildren(VdlGridTile), 
        __metadata('design:type', QueryList)
    ], VdlGridList.prototype, "_tiles", void 0);
    __decorate([
        Input(), 
        __metadata('design:type', Object)
    ], VdlGridList.prototype, "cols", null);
    __decorate([
        Input(), 
        __metadata('design:type', Object)
    ], VdlGridList.prototype, "gutterSize", null);
    __decorate([
        Input(), 
        __metadata('design:type', Object), 
        __metadata('design:paramtypes', [Object])
    ], VdlGridList.prototype, "rowHeight", null);
    VdlGridList = __decorate([
        Component({selector: 'vdl-grid-list',
            template: "<div><ng-content></ng-content></div>",
            styles: [".vdl-grid-list{display:block;position:relative}.vdl-grid-tile{display:block;position:absolute;overflow:hidden}.vdl-grid-tile .vdl-figure{display:flex;position:absolute;align-items:center;justify-content:center;height:100%;top:0;right:0;bottom:0;left:0;padding:0;margin:0}.vdl-grid-tile .vdl-grid-tile-footer,.vdl-grid-tile .vdl-grid-tile-header{display:flex;align-items:center;height:48px;color:#fff;background:rgba(0,0,0,.38);overflow:hidden;padding:0 16px;font-size:16px;position:absolute;left:0;right:0}.vdl-grid-tile .vdl-grid-tile-footer [vdl-line],.vdl-grid-tile .vdl-grid-tile-header [vdl-line]{white-space:nowrap;overflow-x:hidden;text-overflow:ellipsis;display:block;box-sizing:border-box}.vdl-grid-tile .vdl-grid-tile-footer [vdl-line]:nth-child(n+2),.vdl-grid-tile .vdl-grid-tile-header [vdl-line]:nth-child(n+2){font-size:12px}.vdl-grid-tile .vdl-grid-tile-footer>*,.vdl-grid-tile .vdl-grid-tile-header>*{margin:0;padding:0;font-weight:400;font-size:inherit}.vdl-grid-tile .vdl-grid-tile-footer.vdl-2-line,.vdl-grid-tile .vdl-grid-tile-header.vdl-2-line{height:68px}.vdl-grid-tile .vdl-grid-list-text{display:flex;flex-direction:column;width:100%;box-sizing:border-box;overflow:hidden}.vdl-grid-tile .vdl-grid-list-text>*{margin:0;padding:0;font-weight:400;font-size:inherit}.vdl-grid-tile .vdl-grid-list-text:empty{display:none}.vdl-grid-tile .vdl-grid-tile-header{top:0}.vdl-grid-tile .vdl-grid-tile-footer{bottom:0}.vdl-grid-tile .vdl-grid-avatar{padding-right:16px}[dir=rtl] .vdl-grid-tile .vdl-grid-avatar{padding-right:0;padding-left:16px}.vdl-grid-tile .vdl-grid-avatar:empty{display:none}"],
            host: {
                'role': 'list',
                '[class.vdl-grid-list]': 'true',
            },
            encapsulation: ViewEncapsulation.None,
        }),
        __param(2, Optional()), 
        __metadata('design:paramtypes', [Renderer, ElementRef, Dir])
    ], VdlGridList);
    return VdlGridList;
}());
export var VdlGridListModule = (function () {
    function VdlGridListModule() {
    }
    /** @deprecated */
    VdlGridListModule.forRoot = function () {
        return {
            ngModule: VdlGridListModule,
            providers: []
        };
    };
    VdlGridListModule = __decorate([
        NgModule({
            imports: [VdlLineModule],
            exports: [
                VdlGridList,
                VdlGridTile,
                VdlGridTileText,
                VdlLineModule,
                VdlGridTileHeaderCssMatStyler,
                VdlGridTileFooterCssMatStyler,
                VdlGridAvatarCssMatStyler
            ],
            declarations: [
                VdlGridList,
                VdlGridTile,
                VdlGridTileText,
                VdlGridTileHeaderCssMatStyler,
                VdlGridTileFooterCssMatStyler,
                VdlGridAvatarCssMatStyler],
        }), 
        __metadata('design:paramtypes', [])
    ], VdlGridListModule);
    return VdlGridListModule;
}());
